### PR TITLE
Expose Sequence Number based Optimistic Concurrency Control in the rest layer

### PR DESCRIPTION
**Pull request by bleskes**
Relates #36148 
Relates #10708

NB: there are some naming discussions about parameters that resulted in the need to rename the parameters. I will do so in a follow up PR.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/rest/action/document/RestDeleteAction.java [hunk 1]
```diff
@@ -66,6 +67,8 @@ public RestChannelConsumer prepareRequest(final RestRequest request, final NodeC
         deleteRequest.setRefreshPolicy(request.param("refresh"));
         deleteRequest.version(RestActions.parseVersion(request));
         deleteRequest.versionType(VersionType.fromString(request.param("version_type"), deleteRequest.versionType()));
+        deleteRequest.setIfMatch(
+            request.paramAsLong("if_seq_no_match", SequenceNumbers.UNASSIGNED_SEQ_NO), request.paramAsLong("if_primary_term_match", 0));
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/rest/action/document/RestDeleteAction.java]: can you use the values of the getters (`deleteRequest.ifSeqNoMatch()` and `deleteRequest.ifPrimaryTermMatch()`) as fallbacks. This ensures that the defaults on the REST layer agree with those on the transport layer.
See how it's done for the versionType one line above this.

File: server/src/main/java/org/elasticsearch/rest/action/document/RestIndexAction.java [hunk 2]
```diff
@@ -93,6 +94,8 @@ public RestChannelConsumer prepareRequest(final RestRequest request, final NodeC
         indexRequest.setRefreshPolicy(request.param("refresh"));
         indexRequest.version(RestActions.parseVersion(request));
         indexRequest.versionType(VersionType.fromString(request.param("version_type"), indexRequest.versionType()));
+        indexRequest.ifMatch(
+            request.paramAsLong("if_seq_no_match", SequenceNumbers.UNASSIGNED_SEQ_NO), request.paramAsLong("if_primary_term_match", 0));
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/rest/action/document/RestIndexAction.java]: same here

**Review by ywelsch — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/rest/action/document/RestDeleteAction.java (hunk 1 above)
Comment (bleskes) [reply server/src/main/java/org/elasticsearch/rest/action/document/RestDeleteAction.java]: +1

**Comment by bleskes** [comment]
run default distro tests

**Pull request merged**